src/hardware/serialport/serialmouse.cpp

```diff
@@ -0,0 +1,453 @@
+            packet_len = 4;
+        } else {
+            packet_len = 3;           
+        }
+        xmit_2part = false;
+
+    } else if (mouse_type == MouseType::MOUSE_SYSTEMS) {
+        //          -- -- -- -- -- -- -- --
+        // Byte 0:   1  0  0  0  0 LB MB RB
+        // Byte 1:  X7 X6 X5 X4 X3 X2 X1 X0
+        // Byte 2:  Y7 Y6 Y5 Y4 Y3 Y2 Y1 Y0
+
+        uint8_t dx = static_cast<uint8_t>(std::clamp(mouse_delta_x,  -0x80, 0x7f));
+        uint8_t dy = static_cast<uint8_t>(std::clamp(-mouse_delta_y, -0x80, 0x7f));
+        uint8_t bt = mouse_has_3rd_button ? ((~mouse_buttons) & 7) : ((~mouse_buttons) & 3);
+
+        packet[0]  = static_cast<uint8_t>(0x80 | ((bt & 1) << 2) | ((bt & 2) >> 1) | ((bt & 4) >> 1));
+        packet[1]  = dx;
+        packet[2]  = dy;
+        packet_len = 3;
+        xmit_2part = true; // next part contains mouse movement since the start of the 1st part
+
+    } else {
+        unimplemented();
+    }
+
+    clearCounters();
+
+    // send packet
+    xmit_idx            = 0;
+    xmit_another_button = false;
+    xmit_another_move   = false;
+    setEvent(SERIAL_RX_EVENT, bytetime / smooth_div);
+}
+
+void CSerialMouse::startPacketPart2() {
+    // port settings are valid at this point
+
+    if (mouse_type == MouseType::MOUSE_SYSTEMS) {
+        //          -- -- -- -- -- -- -- --
+        // Byte 3:  X7 X6 X5 X4 X3 X2 X1 X0
+        // Byte 4:  Y7 Y6 Y5 Y4 Y3 Y2 Y1 Y0
+
+        uint8_t dx = static_cast<uint8_t>(std::clamp(mouse_delta_x,  -0x80, 0x7f));
+        uint8_t dy = static_cast<uint8_t>(std::clamp(-mouse_delta_y, -0x80, 0x7f));
+
+        packet[0]  = dx;
+        packet[1]  = dy;
+
+        packet_len = 2;
+        xmit_2part = false;
+    } else {
+        unimplemented();
+    }
+
+    clearCounters();
+
+    // send packet
+    xmit_idx            = 0;
+    xmit_another_move   = false;
+    setEvent(SERIAL_RX_EVENT, bytetime / smooth_div);
+}
+
+void CSerialMouse::unimplemented() {
+    LOG_ERR("MOUSE (COM%d): Missing implementation", port_num);
+}
+
+void CSerialMouse::handleUpperEvent(uint16_t type) {
+    if (type == SERIAL_TX_EVENT) {
+        ByteTransmitted(); // tx timeout
+    } else if (type == SERIAL_THR_EVENT) {
+        ByteTransmitting();
+        setEvent(SERIAL_TX_EVENT, bytetime / smooth_div);
+    }
+    else if (type == SERIAL_RX_EVENT) {
+        // check for bytes to be sent to port
+        if (CSerial::CanReceiveByte()) {
+            if (send_ack) {
+                send_ack = false;
+                startPacketId();
+            } else if (xmit_idx < packet_len) {
+                CSerial::receiveByte(packet[xmit_idx++]);
+                if (xmit_idx >= packet_len && xmit_2part)
+                    startPacketPart2();
+                else if (xmit_idx >= packet_len && (xmit_another_move || xmit_another_button))
+                    startPacketData();
+                else
+                    setEvent(SERIAL_RX_EVENT, bytetime / smooth_div);
```

Comment:
This exact call with the division repeated 7 times, it would be better to be encapsulated into a function for consistency.